Introduce ROS interface abstraction to improve testability

anhosi opened the pull request:
The free functions in the ros_integration namespace cannot be mocked in tests. This PR introduces two interfaces, a generic ros client interface (init, shutdown, ok) and a ros node interface for all the mode-centric functions.

anhosi committed:
Introduce interface for ros abstraction

- move all public free function of ros_integration namespace to
  interface
- add test for correct use of ros shutdown in VisualizerApp

anhosi committed:
Document recommendation for testable code

anhosi committed:
Move free functions from rclcpp_node_storage into RosAbstraction class

anhosi committed:
Rename RosAbstraction to RosClientAbstraction

anhosi committed:
Introduce RosNodeAbstraction for the node-centric part of the ros interface

- also replaces the free function in get_topic_names_and_types

anhosi committed:
Cleanup: remove default arguments for virtual function

anhosi committed:
Refactor towards proper usage of RosNodeAbstraction

wjwwood committed:
nitpick: alphabetical ordering

wjwwood commented on rviz_common/include/rviz_common/ros_integration/ros_node_abstraction.hpp:
```diff
@@ -0,0 +1,106 @@
+namespace ros_integration
+{
+
+class RosNodeAbstraction : public RosNodeAbstractionIface
```
I think this class is conflating two things, storage of nodes and the abstraction of nodes. Ideally, the storage of the nodes would be not public, whereas the abstraction for nodes would be public and would also not reference ROS 2 API directly (which the storage does since it takes and returns `rclcpp::Node`'s).

wjwwood commented on rviz_common/include/rviz_common/ros_integration/ros_node_abstraction_iface.hpp:
```diff
@@ -0,0 +1,102 @@
+   *
+   * \param node_name name to be used as the key for the rclcpp node
+   * \param node the rclcpp node to be stored
+   */
```
Probably best to not duplicate the documentation, instead pick one place and document it there. I would guess the non-Iface class is best so long as it is public. It's possible you'd want to document both, but the content should be materially different, otherwise one should just reference the other.

wjwwood commented on rviz_common/src/rviz_common/add_display_dialog.cpp:
```diff
@@ -446,8 +452,10 @@ void DisplayTypeTree::fillTree(Factory * factory)
+TopicDisplayWidget::TopicDisplayWidget(
+  const std::string & node_name,
+  std::unique_ptr<rviz_common::ros_integration::RosNodeAbstractionIface> node_abstraction)
+: node_name_(node_name), node_abstraction_(std::move(node_abstraction))
```
If you're going to pass in a ros node abstraction, just get the name from that rather than also passing the node name along and storing it.

wjwwood commented on rviz_common/include/rviz_common/visualizer_app.hpp:
```diff
@@ -70,6 +74,7 @@ private Q_SLOTS:
   QTimer * continue_timer_;
   VisualizationFrame * frame_;
   std::string node_name_;
+  std::unique_ptr<rviz_common::ros_integration::RosClientAbstractionIface> ros_client_;
```
nitpick: this name `ros_client_` leaves out "abstraction", but in the other file you name the variable `node_abstraction` leaving our "ros". I don't mind leaving out ros or abstraction (though my default preference would be to include both in both variable names), but it would be best to keep them at least consistent, e.g. `client_abstraction_` or `ros_node_`.

wjwwood reviewed (changes requested):
I have some requested changes, but I agree with the direction:

- mock-able abstraction classes which inherit from pure virtual interfaces > free functions
- dependency injection of mock-able abstraction classes
- using interfaces in the code, rather than the child class
- using `unique_ptr`'s for storage that automatically cleans up

anhosi committed:
Comment only non interface classes (avoid duplicate documentation)

anhosi committed:
Cleanup: use consistent variable naming

anhosi committed:
Extract storage functionality from RosNodeAbstraction

anhosi committed:
Refactor RosNodeAbstraction towards a replacement of rclcpp::Node

- a RosNodeAbstraction shall encapsulate a rclcpp::Node in the future
  and thus keeps track of its name (this simplifies a few functions)
- for testing the rest of rviz the public api from the ros_integration
  namespace should be mocked
- the optional storage parameter of RosNodeAbstraction is intended for
  unit testing of RosNodeAbstraction only and should not be needed
  otherwise

anhosi commented:
First try with the ros ci:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=3800)](http://ci.ros2.org/job/ci_linux/3800/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=936)](http://ci.ros2.org/job/ci_linux-aarch64/936/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3138)](http://ci.ros2.org/job/ci_osx/3138/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=3886)](http://ci.ros2.org/job/ci_windows/3886/) (still disabled)

anhosi commented:
The ci build failures are due to the `rosidl_generator_py` package :-(

anhosi commented:
Next ci try (ros2 master is green again)
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=3802)](http://ci.ros2.org/job/ci_linux/3802/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=939)](http://ci.ros2.org/job/ci_linux-aarch64/939/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3141)](http://ci.ros2.org/job/ci_osx/3141/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=3889)](http://ci.ros2.org/job/ci_windows/3889/) (rviz still disabled)

anhosi committed:
Fix gmock dependency

anhosi commented:
Next try with ci :smile: 
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=3805)](http://ci.ros2.org/job/ci_linux/3805/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=943)](http://ci.ros2.org/job/ci_linux-aarch64/943/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3144)](http://ci.ros2.org/job/ci_osx/3144/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=3892)](http://ci.ros2.org/job/ci_windows/3892/)

wjwwood commented on rviz_common/include/rviz_common/ros_integration/ros_client_abstraction.hpp:
```diff
@@ -0,0 +1,102 @@
+   * \return name of the resulting ROS node
+   */
+  std::string
+  init(int argc, char ** argv, const std::string & name, bool anonymous_name) override;
```
Should this return a `RosNodeAbstraction` instead?

anhosi replied:
:+1:  That would be a nicer interface.

wjwwood replied:
Will you do that now or should I go ahead and merge this and you can change it later/follow up pr?

anhosi replied:
I will do that now. It should be a small change.

anhosi replied:
This is more involved than I first thought. Currently we use the node name as a key to easily get access to an `RosNodeAbstraction` instance. The lifetime of this node abstraction has no effect on the stored `rclcpp::Node` inside the `RosNodeStorage` but maybe should have (?). What should be the behavior is the backing `rclcpp::Node` corresponding to a RosNodeAbstraction is clean up? ...
Any suggestions/ thoughts on that @wjwwood ?

I still agree that returning "some kind" of node from `init` is preferable but I am yet not sure how to do that properly. So I would prefer to do this in a follow up pr.

wjwwood replied:
I'd probably not make the node abstraction object scoped, just because you cannot "cleanup" a ROS 1 node, so maybe we don't need to worry about it.

wjwwood commented on rviz_common/include/rviz_common/ros_integration/ros_node_abstraction.hpp:
```diff
@@ -0,0 +1,102 @@
+   *
+   * \param node_name name of the node to create
+   */
+  explicit RosNodeAbstraction(const std::string & node_name);
```
I sort of think it would be better to have the constructor private and only generate these instances via a `get_or_create_node_by_name()` (or some other name, just trying to be specific here) method of the `RosClientAbstrction` class.

anhosi replied:
Introducing a static factory function does not allow mocking and thus kind of defeats the purpose of these changes. I agree that the current constructor solution is not ideal. I will try to expore using a separate factory.

wjwwood replied:
Ok, well don't spend too much time on it, if you come up with something that works better but still lets you mock it then great, otherwise we'll take it as-is.

The other thing to think about is how do you destroy nodes that have been created? In the case of ROS 2 (at least) you can make more than one node and you can remove them too. Maybe the `RosClientAbstraction` needs a destroy node method which is implicitly called on `shutdown()` if not already called for each node (typically just the one created during `init()`). Or the `RosNodeAbstraction` could be a scoped object (could be a `shared_ptr` with the node storage keeping only a `weak_ptr`). Dunno, just something to think about.

wjwwood reviewed (commented):
lgtm, other two small questions/comments about the new API.

wjwwood reviewed (approved):
questions answered, lgtm with the different return type of `init()` as an optional improvement.
